gcc/rust/rust-session-manager.cc

```diff
@@ -439,20 +439,37 @@ Session::compile_crate (const char *filename)
   // parse file here
   /* create lexer and parser - these are file-specific and so aren't instance
    * variables */
-  Lexer lex (filename, std::move (file_wrap), linemap);
+  Optional<std::ofstream &> dump_lex_opt = Optional<std::ofstream &>::none ();
+  std::ofstream dump_lex_stream;
+  if (options.dump_option_enabled (CompileOptions::LEXER_DUMP))
+    {
+      dump_lex_stream.open (kLexDumpFile);
+      if (dump_lex_stream.fail ())
+	{
+	  rust_error_at (Linemap::unknown_location (),
+			 "cannot open %s:%m; ignored", kLexDumpFile);
+	}
+      auto stream = Optional<std::ofstream &>::some (dump_lex_stream);
+      dump_lex_opt = std::move (stream);
+    }
+
+  Lexer lex (filename, std::move (file_wrap), linemap, dump_lex_opt);
+
   Parser<Lexer> parser (lex);
 
   // generate crate from parser
   std::unique_ptr<AST::Crate> ast_crate = parser.parse_crate ();
 
+  if (dump_lex_opt.is_some ())
+    {
+      std::ofstream &out = dump_lex_opt.get ();
+      out.close ();
+    }
+
```

Comment:
Is this necessary? The `Optional` should simply go out of scope and destroy its member, in that case the `ofstream` whose destructor takes care of closing the stream?

Reply:
Oh you are correct. Destructor of ofstream is automatically called and closes the stream.